### PR TITLE
PHP7 Support

### PR DESCRIPTION
**Pull request by Sean-Der**
Hey @laruence !
I have finally gotten this to the point that I feel confident about the architecture. 

There are still many test fails (pass rate is at around 30%) However, most of these test fails have to do with not properly handling references and other issues that can be fixed easily (I just don't understand the zend engine API well enough)

I am opening this branch so I can mark some questions for you, and start a light review.

Would it be possible to start running some CI against this using the current service?

I am removing all preprocessor stuff to handle different API versions, I am only supporting PHP 7.

### CODE REVIEW

**Comment by laruence** [comment]
Wow, great, anyway, I won't be able to look into this in the next week... (maybe I will start do it in the next weekend).. thanks in advance.

File: msgpack.c [hunk 1]
```diff
@@ -147,81 +147,53 @@ ZEND_GET_MODULE(msgpack)
...
-    }
-
-    smart_str_0(&buf);
-    *newstr = buf.c;
-
+    msgpack_serialize_zval(&buf, &PS(http_session_vars), var_hash);
     msgpack_serialize_var_destroy(&var_hash);
 
-    return SUCCESS;
+    z_string = zend_string_init(buf.c, buf.len, 0);
```
Comment (laruence) [inline msgpack.c]: maybe  msgpack_serialize_zval should use zend_str directly? then we can save a string duplication here

File: msgpack.c [hunk 2]
```diff
@@ -147,81 +147,53 @@ ZEND_GET_MODULE(msgpack)
...
-                    break;
-                case HASH_KEY_IS_STRING:
-                    php_set_session_var(
-                        key_str, key_len - 1, *value, NULL TSRMLS_CC);
-                    php_add_session_var(key_str, key_len - 1 TSRMLS_CC);
-                    break;
+    if (ret == MSGPACK_UNPACK_EXTRA_BYTES || ret == MSGPACK_UNPACK_SUCCESS) {
+        msgpack_unserialize_var_destroy(&var_hash, 0, &tmp);
+
+        ZEND_HASH_FOREACH_KEY_VAL(HASH_OF(&tmp), key_long, key_str, value) {
```
Comment (laruence) [inline msgpack.c]: if you don't care about key_long, then use ZEND_HASH_FOREACH_STR_KEY_VAL here might be better

File: msgpack_class.c [hunk 3]
```diff
@@ -528,16 +455,18 @@ static ZEND_METHOD(msgpack_unpacker, execute)
 
     if (unpacker->retval == NULL)
     {
-        ALLOC_INIT_ZVAL(unpacker->retval);
+        zval zv;
+        unpacker->retval = &zv;
```
Comment (laruence) [inline msgpack_class.c]: reference to a local stack value is undefined and dangerous behavior, you need move zval zv declaration out of this block.

File: msgpack_class.c [hunk 4]
```diff
@@ -602,7 +531,6 @@ static ZEND_METHOD(msgpack_unpacker, data)
         {
             zval *zv;
 
-            ALLOC_INIT_ZVAL(zv);
             ZVAL_ZVAL(zv, unpacker->retval, 1, 0);
```
Comment (laruence) [inline msgpack_class.c]: seems wrong here, zv is a uninitlized zval pointer...

File: msgpack_convert.c [hunk 5]
```diff
@@ -454,16 +341,15 @@ int msgpack_convert_object(zval *return_value, zval *tpl, zval **value)
...
 
-    if (Z_TYPE_PP(value) == IS_OBJECT)
-    {
+    if (Z_TYPE_P(*value) == IS_OBJECT) {
         zend_class_entry *vce;
 
-        vce = zend_get_class_entry(*value TSRMLS_CC);
-        if (strcmp(ce->name, vce->name) == 0)
+        vce = Z_OBJCE_P(tpl);
+        if (strcmp(ce->name->val, vce->name->val) == 0)
```
Comment (laruence) [inline msgpack_convert.c]: use zend_string_equal is better

File: msgpack_unpack.c [hunk 6]
```diff
@@ -694,66 +556,70 @@ int msgpack_unserialize_map_item(
         }
     }
 
-    if (Z_TYPE_PP(container) != IS_ARRAY && Z_TYPE_PP(container) != IS_OBJECT)
-    {
+    zend_bool container_is_ref = 0;
```
Comment (Sean-Der) [inline msgpack_unpack.c]: @laruence 
What is the proper way to ZVAL_DEREF *container here, without actually modifying *container?

Calling array_init on the result of Z_REFVAL_P causes issues, I just want to modify the REFVAL of *container, without harming the ref *container

File: msgpack_unpack.c [hunk 7]
```diff
@@ -694,66 +556,70 @@ int msgpack_unserialize_map_item(
...
-                zval_ptr_dtor(&val);
-                MSGPACK_WARNING(
-                    "[msgpack] (%s) illegal offset type, skip this decoding",
-                    __FUNCTION__);
-            }
-            zval_ptr_dtor(&key);
-            break;
-        default:
-            MSGPACK_WARNING("[msgpack] (%s) illegal key type", __FUNCTION__);
+    if (Z_TYPE_P(*container) == IS_OBJECT && Z_OBJCE_P(*container) != PHP_IC_ENTRY) {
```
Comment (Sean-Der) [inline msgpack_unpack.c]: Calling zend_symtable_update on the HASH_OF a class breaks the class in PHP 7. Later when trying to modify the class you can't access members added via zend_symtable_update, and the class itself becomes immutable.

Is this expected behavior? Was msgpack-php relying on undefined behavior in <7 ?

File: msgpack_pack.c [hunk 8]
```diff
@@ -10,242 +10,172 @@
...
+#define msgpack_pack_user smart_string*
 #define msgpack_pack_inline_func(name) \
     static inline void msgpack_pack ## name
 #define msgpack_pack_inline_func_cint(name) \
     static inline void msgpack_pack ## name
 #define msgpack_pack_append_buffer(user, buf, len) \
-    smart_str_appendl(user, (const void*)buf, len)
+    smart_string_appendl(user, (const void*)buf, len)
 #include "msgpack/pack_template.h"
+#define Z_REF_AWARE_P(_obj) \
```
Comment (Sean-Der) [inline msgpack_pack.c]: This macro feels pointless, but I couldn't find a way to implement this functionality without a lot of extra code

**Comment by Sean-Der** [comment]
@laruence 
I am at 55.0% on test fails, getting there!

However, I am confused about the behavior of printing references in 5 vs 7.

Here is a gist of how PHP 5 would print a zval with a '&' in even if no ADDREF had been called, in 7 ADDREF must be called. https://gist.github.com/Sean-Der/1be1c8b998d95ef51bee

The change in ext/standard/var.c in https://github.com/php/php-src/commit/2eb980f614ab59aa8f632ae5a70f719f34e8410f causes the issue, if I revert this all the ref tests are good again.

**Comment by Sean-Der** [comment]
Hey @laruence, first off good news! We are at 85% with test fails, holy crap I never thought I would see the day.... https://gist.github.com/Sean-Der/cd94c0b6fb71a81e089b

However, I have ran into another issue with translating. In PHP 5 you could call zend_symtable_update and zend_hash_index_update on an object, allowing you to create invalid properties. In PHP 7 when you use those functions on the HASH_OF(object) it will create strange state (object becomes immutable, you will get a segfault in other places)
[Here](https://gist.github.com/Sean-Der/f8cb8578f195dc563414) is the sample code in PHP 5 showing these behaviors.


In PHP 7 converting these functions to the appropriate zend_object functions also breaks references. [zend_std_write_property](http://lxr.php.net/xref/phpng/Zend/zend_object_handlers.c#663)
[Here](https://gist.github.com/Sean-Der/4a07a044dab4dd23e46b) is the sample code in PHP 5 showing these behaviors.

-----------

I am going to be taking a break from working on this extension till I know the answer to the issues I have talked about so far. The template related functions have memory leaks (that are caught by the PHP memory manager)

thanks!

File: msgpack_class.c [hunk 9]
```diff
@@ -492,61 +364,41 @@ static ZEND_METHOD(msgpack_unpacker, execute)
 {
     char *str = NULL, *data;
     long str_len = 0;
```
Comment (zxcvdavid) [inline msgpack_class.c]: str_len should be size_t

File: msgpack_class.c [hunk 10]
```diff
@@ -492,61 +364,41 @@ static ZEND_METHOD(msgpack_unpacker, execute)
...
-    if (str != NULL)
-    {
-        data = (char *)str;
-        len = (size_t)str_len;
-        if (offset != NULL)
-        {
+    if (str) {
+        data = str;
+        len = str_len;
+        if (offset != NULL) {
```
Comment (zxcvdavid) [inline msgpack_class.c]: need check offset type

File: msgpack_class.c [hunk 11]
```diff
@@ -583,54 +430,49 @@ static ZEND_METHOD(msgpack_unpacker, execute)
...
+            RETURN_NULL();
+        }
     }
 
+    ZVAL_STRING(&func_name, "reset");
+    call_user_function_ex(CG(function_table), getThis(), &func_name, &reset_return, 0, NULL, 0, NULL);
+    zval_ptr_dtor(&func_name);
+
+    return;
     RETURN_FALSE;
```
Comment (zxcvdavid) [inline msgpack_class.c]: dead code

File: msgpack_convert.c [hunk 12]
```diff
@@ -5,448 +5,319 @@
...
+            } else {
                 break;
             }
         }
-
         *properties = NULL;
     }
-
-    return zend_hash_index_update(ht, key_index, &val, sizeof(val), NULL);
+    zval key_zv;
```
Comment (zxcvdavid) [inline msgpack_convert.c]: C89 issue

File: msgpack_convert.c [hunk 13]
```diff
@@ -5,448 +5,319 @@
...
+    ZVAL_LONG(&key_zv, key_index);
+    zend_std_write_property(object, &key_zv, val, NULL);
+    return SUCCESS;
 }
 
-static inline int msgpack_convert_string_to_properties(
+static inline int  msgpack_convert_string_to_properties(
```
Comment (zxcvdavid) [inline msgpack_convert.c]: delete the space

File: msgpack_convert.c [hunk 14]
```diff
@@ -5,448 +5,319 @@
...
-        if (num <= 0)
-        {
-            MSGPACK_WARNING(
-                "[msgpack] (%s) template array length is 0",
+    zend_string *key;
+    int key_type;
+    ulong key_index;
+    zval *data, *arydata;
+    HashPosition pos, valpos;
+    HashTable *ht, *htval;
```
Comment (zxcvdavid) [inline msgpack_convert.c]: C89 issue

File: msgpack_convert.c [hunk 15]
```diff
@@ -471,228 +340,112 @@ int msgpack_convert_object(zval *return_value, zval *tpl, zval **value)
...
-                        if (msgpack_convert_string_to_properties(
-                                return_value, key, key_len, val, var) != SUCCESS)
-                        {
-                            zval_ptr_dtor(&val);
-                            MSGPACK_WARNING(
-                                "[msgpack] (%s) "
-                                "illegal offset type, skip this decoding",
-                                __FUNCTION__);
+                ZEND_HASH_FOREACH_STR_KEY_VAL(ht, str_key, data) {
+                    if (str_key) {
```
Comment (zxcvdavid) [inline msgpack_convert.c]: it is not necessary

**Comment by Sean-Der** [comment]
@zxcvdavid Thank you so much for looking this PR over! Going through and fixing your comments now.

I am not able to compile with -C89 , I am getting errors in Zend headers? What compiler/flags are recommended?

Would you mind looking over my comments/questions above about the zend API.

thanks!

File: msgpack_convert.c (hunk 15 above)
Comment (Sean-Der) [inline msgpack_convert.c]: I *think* ZEND_HASH_FOREACH_STR_KEY_VAL will return data that is a numeric key and str_key will just be null.

ZEND_HASH_FOREACH_STR_KEY_VAL just sets up pointers and ZEND_HASH_FOREACH isn't doing any filtering.

Comment (zxcvdavid) [inline msgpack_convert.c]: Ohhh.Yes, you are right. ```if (str_key) {``` is necessary. :)

**Comment by zxcvdavid** [comment]
@Sean-Der about C89 issues, you can see : https://github.com/php/php-src/commit/76a290d0be25a028f968edff074c9d8b6c673dba

thanks.

**Comment by laruence** [comment]
Hey, just FYI, I am going to work on this in this week ,  I am going to create a new PHP7 branch.. thanks

**Comment by Sean-Der** [comment]
@laruence that is awesome to hear!

Is there anything I can do to make things easier on you?

**Comment by laruence** [comment]
@Sean-Der nothing now,  you have done a great work :), thanks

**Comment by laruence** [comment]
just for the record,  I spent whole day on this. there is a problem that is, we should pop values on stack. thus in case of err, we could release these garbage on the stack to fix the memleaks. 

however,  it is not easy for the current implementation.. need more time..

my current work could be found at : https://github.com/msgpack/msgpack-php/tree/php7

**Comment by Sean-Der** [comment]
@laruence Just looking at all the things you fixed up, I made SO many mistakes thank you for cleaning up after me :/ doesn't look like I saved you that much time in the end urgh, better luck next time hopefully!

sorry I couldn't figure out a better way to keep the stack an array of zval* , I am excited to see what the proper thing to do was!

thank you so much for putting so much time into this.

**Comment by laruence** [comment]
nah, you did a great progress.  just maybe because I am more familiar with PHP7 internal. :)
anyway, are you still willing to participated in maintaining this ext?

**Comment by laruence** [comment]
@Sean-Der I fixed the memleaks, now I am going to close this PR.. thanks for all the thing you did :)

**Comment by Sean-Der** [comment]
@laruence working with you has been the best experience I have had with Open Source in a very long time. I can't thank you enough for how supporting/nice you have been. You have been awesome, if I run into you a conference I owe you a soda/beer etc...

I would love to continue helping with the extension! (Is that the most helpful thing I could be doing for PHP right now?) 

If so I will work on getting bin/str support in 5+7 and take a look at all existing bugs and triage them. Otherwise, just tell me where I can be most useful! I am currently porting pecl-yaml and in contact with the maintainer, and things are going MUCH faster.

**Comment by laruence** [comment]
@Sean-Der great, I am going to invite you into this project :) thanks

**Pull request closed without merging**